internal/generator/vector/output/otlp/transform.go

```diff
@@ -0,0 +1,339 @@
+package otlp
+
+import (
+	. "github.com/openshift/cluster-logging-operator/internal/generator/framework"
+	"github.com/openshift/cluster-logging-operator/internal/generator/vector/elements"
+	"github.com/openshift/cluster-logging-operator/internal/generator/vector/helpers"
+	"strings"
+)
+
+func TransformContainer(id string, inputs []string) Element {
+	return elements.Remap{
+		Desc:        "Normalize container log records to OTLP semantic conventions",
+		ComponentID: id,
+		Inputs:      helpers.MakeInputs(inputs...),
+		VRL: strings.TrimSpace(`
+# OTLP semantic conventions for application and infrastructure containers
+# https://opentelemetry.io/docs/specs/semconv/resource/k8s/
+resource.attributes = []
+meta = [
+  "kubernetes.pod_name", 
+  "kubernetes.pod_id",
+  "kubernetes.namespace_name",
+  "kubernetes.container_name",
+  "openshift.cluster_id",
+  "hostname",
+]
+replacements = {
+  "pod.id": "pod.uid",
+  "cluster.id": "cluster.uid",
+  "hostname": "node.name",
+}
+for_each(meta) -> |_,sub_key| {
+  path = split(sub_key,".")
+  if length(path) > 1 {
+	sub_key = replace!(path[-1],"_",".")
+  }
+  if get!(replacements, [sub_key]) != null {
+	sub_key = string!(get!(replacements, [sub_key]))
+  } 
+  resource.attributes = append(
+	resource.attributes,
+	[{"key": "k8s." + sub_key, "value": {"stringValue": get!(.,path)}}]
+  )
+}
+# Append kube pod labels
+if exists(.kubernetes.labels) {
+	for_each(object!(.kubernetes.labels)) -> |key,value| {  
+		resource.attributes = append(
+			resource.attributes,
+			[{"key": "k8s.pod.label." + key, "value": {"stringValue": value}}]
+		)
+	}
+}
+# Append container data
+container_meta = [
+  "kubernetes.container_name",
+  "kubernetes.container_id",
+  "kubernetes.container_image",
+]
+for_each(container_meta) -> |_,sub_key| {
+  path = split(sub_key,".")
+  if length(path) > 1 {
+	sub_key = replace!(path[-1],"_",".")
+  }
+  resource.attributes = append(
+	resource.attributes,
+	[{"key": "container." + sub_key, "value": {"stringValue": get!(.,path)}}]
+  )
+}
+
+# Transform the record
```

Comment:
Maybe not first pass but there is opportunity to break a number of these into smaller, reusable bits to be composed into the transform we need

Reply:
Very much agreed

Reply:
done